### ISSUE TITLE
git master: segfault

### ISSUE THREAD

**Issue by f2404** [gnunn1/terminix#745]
```
Feb 06 18:27:26 kernel: terminix[12575]: segfault at 20 ip 0000000000557bf0 sp 00007ffe90ce6940 error 4 in terminix[400000+909000]
Feb 06 18:27:26 systemd[1]: Started Process Core Dump (PID 12961/UID 0).
Feb 06 18:27:27 systemd-coredump[12962]: Process 12575 (terminix) of user 1000 dumped core.
                                         
                                         Stack trace of thread 12575:
                                         #0  0x0000000000557bf0 _D2gx8terminix11customtitle11CustomTitle13removeTimeoutMFZv (terminix)
                                         #1  0x00000000005577d7 _D2gx8terminix11customtitle11CustomTitle13onButtonPressMFC3gdk5Event5EventC3gtk6Widget6WidgetZb (terminix)
                                         #2  0x000000000078fce1 _D3gtk6Widget6Widget31callBackButtonPressEventGenericUPS4gtkc8gtktypes9GtkWidgetPS4gtkc8gdktypes8GdkEventC3gtk6Widget6Widget40OnButtonPressEventGenericDelegateWrapperZi (terminix)
                                         #3  0x00007f5a393cf96c n/a (libgtk-3.so.0)
                                         #4  0x00007f5a3ce3af75 g_closure_invoke (libgobject-2.0.so.0)
                                         #5  0x00007f5a3ce4cf82 n/a (libgobject-2.0.so.0)
                                         #6  0x00007f5a3ce5566f g_signal_emit_valist (libgobject-2.0.so.0)
                                         #7  0x00007f5a3ce55faf g_signal_emit (libgobject-2.0.so.0)
                                         #8  0x00007f5a3951c31c n/a (libgtk-3.so.0)
                                         #9  0x00007f5a393cca3e n/a (libgtk-3.so.0)
                                         #10 0x00007f5a393ceaae gtk_main_do_event (libgtk-3.so.0)
                                         #11 0x00007f5a3bd00d15 n/a (libgdk-3.so.0)
                                         #12 0x00007f5a3bd31d42 n/a (libgdk-3.so.0)
                                         #13 0x00007f5a41cf3587 g_main_context_dispatch (libglib-2.0.so.0)
                                         #14 0x00007f5a41cf37f0 n/a (libglib-2.0.so.0)
                                         #15 0x00007f5a41cf389c g_main_context_iteration (libglib-2.0.so.0)
                                         #16 0x00007f5a3c90f54d g_application_run (libgio-2.0.so.0)
                                         #17 0x00000000005fd28a _D3gio11Application11Application3runMFAAyaZi (terminix)
                                         #18 0x000000000047d4a6 _Dmain (terminix)
                                         #19 0x00007f5a4337595f _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv (libdruntime-ldc.so.70)
                                         #20 0x00007f5a433758a4 _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ7tryExecMFMDFZvZv (libdruntime-ldc.so.70)
                                         #21 0x00007f5a433757cd _d_run_main (libdruntime-ldc.so.70)
                                         #22 0x00000000004ef1b8 main (terminix)
                                         #23 0x00007f5a42826291 __libc_start_main (libc.so.6)
                                         #24 0x000000000047c4ca _start (terminix)
                                         
                                         Stack trace of thread 12576:
                                         #0  0x00007f5a428e548d poll (libc.so.6)
                                         #1  0x00007f5a41cf3786 n/a (libglib-2.0.so.0)
                                         #2  0x00007f5a41cf389c g_main_context_iteration (libglib-2.0.so.0)
                                         #3  0x00007f5a41cf38e1 n/a (libglib-2.0.so.0)
                                         #4  0x00007f5a41d1b0d5 n/a (libglib-2.0.so.0)
                                         #5  0x00007f5a42eaf454 start_thread (libpthread.so.0)
                                         #6  0x00007f5a428ee7df __clone (libc.so.6)
                                         
                                         Stack trace of thread 12577:
                                         #0  0x00007f5a428e548d poll (libc.so.6)
                                         #1  0x00007f5a41cf3786 n/a (libglib-2.0.so.0)
                                         #2  0x00007f5a41cf3b12 g_main_loop_run (libglib-2.0.so.0)
                                         #3  0x00007f5a3c93b316 n/a (libgio-2.0.so.0)
                                         #4  0x00007f5a41d1b0d5 n/a (libglib-2.0.so.0)
                                         #5  0x00007f5a42eaf454 start_thread (libpthread.so.0)
                                         #6  0x00007f5a428ee7df __clone (libc.so.6)
                                         
                                         Stack trace of thread 12655:
                                         #0  0x00007f5a428e9f19 syscall (libc.so.6)
                                         #1  0x00007f5a41d3903a g_cond_wait_until (libglib-2.0.so.0)
                                         #2  0x00007f5a41cc7e89 n/a (libglib-2.0.so.0)
                                         #3  0x00007f5a41cc84ac g_async_queue_timeout_pop (libglib-2.0.so.0)
                                         #4  0x00007f5a41d1bb9d n/a (libglib-2.0.so.0)
                                         #5  0x00007f5a41d1b0d5 n/a (libglib-2.0.so.0)
                                         #6  0x00007f5a42eaf454 start_thread (libpthread.so.0)
                                         #7  0x00007f5a428ee7df __clone (libc.so.6)
                                         
                                         Stack trace of thread 12578:
                                         #0  0x00007f5a428e548d poll (libc.so.6)
                                         #1  0x00007f5a41cf3786 n/a (libglib-2.0.so.0)
                                         #2  0x00007f5a41cf389c g_main_context_iteration (libglib-2.0.so.0)
                                         #3  0x00007f5a340244bd n/a (libdconfsettings.so)
                                         #4  0x00007f5a41d1b0d5 n/a (libglib-2.0.so.0)
                                         #5  0x00007f5a42eaf454 start_thread (libpthread.so.0)
                                         #6  0x00007f5a428ee7df __clone (libc.so.6)
```

**Comment by gnunn1** [comment]
I took a stab at fixing this based on the stacktrace. Is it reproducible for you and if so could you try the commit and see if it helps?

**Comment by f2404** [comment]
I could reproduce the very same segfault but I've got 2 other with slightly different trace:
```
Feb 06 18:47:06 systemd-coredump[21531]: Process 20856 (terminix) of user 1000 dumped core.
                                         
                                         Stack trace of thread 20856:
                                         #0  0x0000000000657536 _D4glib7Timeout7Timeout16callAllListenersMFZb (terminix)
                                         #1  0x0000000000656e4e _D4glib7Timeout7Timeout15timeoutCallbackUC4glib7Timeout7TimeoutZb (terminix)
                                         #2  0x00007f0fd2c4aeb3 n/a (libglib-2.0.so.0)
                                         #3  0x00007f0fd2c4a43a g_main_context_dispatch (libglib-2.0.so.0)
                                         #4  0x00007f0fd2c4a7f0 n/a (libglib-2.0.so.0)
                                         #5  0x00007f0fd2c4a89c g_main_context_iteration (libglib-2.0.so.0)
                                         #6  0x00007f0fcd86654d g_application_run (libgio-2.0.so.0)
                                         #7  0x00000000005fd28a _D3gio11Application11Application3runMFAAyaZi (terminix)
                                         #8  0x000000000047d4a6 _Dmain (terminix)
                                         #9  0x00007f0fd42cc95f _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv (libdruntime-ldc.so.70)
                                         #10 0x00007f0fd42cc8a4 _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ7tryExecMFMDFZvZv (libdruntime-ldc.so.70)
                                         #11 0x00007f0fd42cc7cd _d_run_main (libdruntime-ldc.so.70)
                                         #12 0x00000000004ef1b8 main (terminix)
                                         #13 0x00007f0fd377d291 __libc_start_main (libc.so.6)
                                         #14 0x000000000047c4ca _start (terminix)
```

**Comment by gnunn1** [comment]
Can you tell me the steps to reproduce it?

**Comment by f2404** [comment]
I was randomly clicking on the header bar (to trigger session name field) while adding/closing terminals and opening and closing new windows and sessions. The exact scenario is unclear to me.

**Comment by f2404** [comment]
I was able to reproduce the 2nd segfault with your latest commit.

**Comment by f2404** [comment]
Sometimes (not always) it crashes in the following scenario:
- open terminix and click header bar to open a field;
- open another terminal down;
- click header bar to open a field;
- close current terminal while the name field is open;
- wait a sec and open Preferences -> segfault.

**Comment by gnunn1** [comment]
I'm not able to reproduce this, however after looking at the code I've noticed one oddity where I was manually setting the timeoutId to 0. This should no longer be necessary after some fixes were made in GtkD and maybe it was causing an issue of some sort. Anyway, try it out and see if you can still reproduce.

**Comment by f2404** [comment]
Nope, it's still there, the same scenario as described above.

**Comment by gnunn1** [comment]
OK I can reproduce with LDC but not DMD, investigating...

**Comment by gnunn1** [comment]
I've tracked this down, it looks like an LDC optimization issue to me. Timeout.d from GtkD has the following code in it:

```
	private bool callAllListeners()
	{
		bool runAgain = false;

		int i = 0;

		while ( i<timeoutListeners.length )
		{
			if ( !timeoutListeners[i]() )
			{
				timeoutListeners = timeoutListeners[0..i] ~ timeoutListeners[i+1..timeoutListeners.length];
			}
			else
			{
				runAgain = true;
				++i;
			}
		}

		// Set timeoutID to 0 if all delegates are removed
		if (timeoutListeners.length == 0) {
			timeoutID = 0;
		}
		return runAgain;
	}
```
What this code does is call the listeners (which is terminix) and return true/false depending on whether any of the listeners want the timer to continue. In my case, the listener always returns false which should result in runAgain returning false. However I can see the timeout continues to be invoked by GTK which indicates it got true as a return value.

My thinking is that LDC is erroneously optimizing runAgain to return true. If I modify the function to add one line below runAgain as follows, it works as expected:

```
	private bool callAllListeners()
	{
		bool runAgain = false;

		int i = 0;

		while ( i<timeoutListeners.length )
		{
			if ( !timeoutListeners[i]() )
			{
				timeoutListeners = timeoutListeners[0..i] ~ timeoutListeners[i+1..timeoutListeners.length];
			}
			else
			{
				runAgain = true;
				++i;
			}
		}

		// Set timeoutID to 0 if all delegates are removed
		if (timeoutListeners.length == 0) {
			timeoutID = 0;
		}
		if (!runAgain) writeln("Run again is false");
		return runAgain;
	}
```

I'll follow-up with the LDC folks and get their thoughts. I did try compiling it with optimization disabled and it made no difference.

**Comment by f2404** [comment]
@gnunn1 Wow, looks like quite an investigation!

**Comment by gnunn1** [comment]
I've fixed this by manually creating the timeout myself instead of relying on the GtkD object for it.

**Comment by f2404** [comment]
Yes, seems to be fixed now, thanks!

**Comment by Nnidyu** [comment]
@gnunn1 I have probably related issue on 1.5.0 (on Fedora 25) and I am not sure if this is the same problem or something separate:

```
Core was generated by `terminix --quake'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00000000005d77ca in glib.Timeout.Timeout.callAllListeners() ()
[Current thread is 1 (Thread 0x7f5f7cb7e740 (LWP 6780))]
Missing separate debuginfos, use: dnf debuginfo-install terminix-1.5.0-1.fc25.x86_64
(gdb) bt
#0  0x00000000005d77ca in glib.Timeout.Timeout.callAllListeners() ()
#1  0x00000000005d714e in glib.Timeout.Timeout.timeoutCallback(glib.Timeout.Timeout) ()
#2  0x00007f5f7b04b89d in g_timeout_dispatch () from /lib64/libglib-2.0.so.0
#3  0x00007f5f7b04ae52 in g_main_context_dispatch () from /lib64/libglib-2.0.so.0
#4  0x00007f5f7b04b1d0 in g_main_context_iterate.isra () from /lib64/libglib-2.0.so.0
#5  0x00007f5f7b04b27c in g_main_context_iteration () from /lib64/libglib-2.0.so.0
#6  0x00007f5f7a149b9d in g_application_run () from /lib64/libgio-2.0.so.0
#7  0x000000000057eecd in gio.Application.Application.run(immutable(char)[][]) ()
#8  0x00000000004ef908 in D main ()
```

**Comment by gnunn1** [comment]
@Nnidyu It looks like it is related to the timeout, however I thought I fixed it in 1.5.0. If you have time, I've posted a debug version of terminix in #803, please give it a try following the instructions and post the backtrace.

**Comment by Nnidyu** [comment]
@gnunn1 Thanks, it seems that it is the same as #803 as I got the same stacktrace now:

```
Thread 1 "terminix" received signal SIGSEGV, Segmentation fault.
0x000000000062a17a in _D4glib7Timeout7Timeout16callAllListenersMFZb (this=0x7fffd4208980) at Timeout.d:192
192	Timeout.d: No such file or directory.
(gdb) bt
#0  0x000000000062a17a in _D4glib7Timeout7Timeout16callAllListenersMFZb (this=0x7fffd4208980) at Timeout.d:192
#1  0x0000000000629afe in _D4glib7Timeout7Timeout15timeoutCallbackUC4glib7Timeout7TimeoutZb (timeout=0x7fffd4208980) at Timeout.d:177
#2  0x00007ffff649389d in g_timeout_dispatch () from /lib64/libglib-2.0.so.0
#3  0x00007ffff6492e52 in g_main_context_dispatch () from /lib64/libglib-2.0.so.0
#4  0x00007ffff64931d0 in g_main_context_iterate.isra () from /lib64/libglib-2.0.so.0
#5  0x00007ffff649327c in g_main_context_iteration () from /lib64/libglib-2.0.so.0
#6  0x00007ffff5591b9d in g_application_run () from /lib64/libgio-2.0.so.0
#7  0x00000000005d187d in _D3gio11Application11Application3runMFAAyaZi (this=0x7ffff7f35800, argv=...) at Application.d:934
#8  0x00000000004f4af9 in _Dmain (args=...) at /home/gnunn/Development/terminix/source/app.d:146
```
